cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/linkanalysis/LinkAnalysis.scala

```diff
@@ -0,0 +1,121 @@
+ * information needed to assess the number of iterations, the error, and the current set of values.
+ * @param error  The T1 error for the current iteration vs the previous iteration.
+ */
+abstract class Iteration(val error: Double, val iteration: Int)
+
+/**
+ * `AbstractLinkAnalysis` is the base class that extends `LinkAnalysis`.  All algorithms should inherit from
+ * this class.  Let it be noted that we exploit scala streams to accomplish the recursion necessary for
+ * our algorithms.  We will not be storing the results every step of the way--instead we find the iteration
+ * that first satisfies the condition of exceeding maximum iterations or falls below the error threshold.
+ * @param params The set of all parameters passed into our algorithm
+ * @param modelName This must be provided by all subclasses for logging purposes.  This should not be
+ *                  left up to the user to define.  This must be provided by the implementer.
+ * @param initial  The starting iteration for our algorithms.  This can be provided by the user or can be
+ *                 given the default value provided by the `start` method.
+ * @tparam T `LinkAnalysis` must be generically typed by `Iteration` or one of its subclasses.  An `Iteration`
+ *           holds all of the pertinent information for a given algorithm.
+ */
+abstract class AbstractLinkAnalysis[T <: Iteration](graph: DirectedGraph[Node],
+    params: Params, modelName: String, initial: Option[T] = None) extends LinkAnalysis[T] {
+
+  protected val log = Logger.get(modelName)
+
+  val maxIterations = params.maxIterations
+  val tolerance     = params.tolerance
+
+  /**
+   * This is the final `Iteration` object from our stream.  The user can feed this iteration back to `iterate`
+   * or `run` if they wish to converge further.
+   */
+  val result = if ((maxIterations.isDefined && maxIterations.get > 0) || !maxIterations.isDefined)
+    this(initial) else start
+
+  /**
+   * The final error associated with the initial run.
+   */
+  val error = result.error
+
+  /**
+   * The final iteration number our algorithm terminated at
+   */
+  val iterations = result.iteration
+
+  def run(init: T = start): T = {
+    lazy val stream: Stream[T] = init #:: stream map { h => iterate(h)}
+
+    // Let the user know if they can save memory!
+    if (graph.maxNodeId.toDouble / graph.nodeCount > 1.1 && graph.maxNodeId - graph.nodeCount > 1000000)
+      log.info("Warning - you may be able to reduce the memory usage by renumbering this graph!")
+
+    log.debug(s"Initializing starting ${modelName}...")
+    val progress = Progress(s"${modelName}_init", 65536, Some(graph.nodeCount))
+
+    stream.find { s =>
+      log.debug("Beginning %sth iteration".format(s.iteration))
+      progress.inc
+      if (maxIterations.isDefined)
+        s.iteration >= maxIterations.get || s.error <= tolerance
+      else
+        s.error <= tolerance
+    }.get
```

Comment:
You're correct.  Using a `stream` caused a large decrease in performance (page rank was taking 3x longer to converge).